grammar/item.g

```diff
@@ -1,7 +1,92 @@
+    Enum:{ "enum" name:IDENT generics:Generics? where_clause:WhereClause? "{" variants:EnumVariant* % "," ","? "}" } |
+    Struct:{ "struct" name:IDENT generics:Generics? body:StructBody } |
+    Union:{ "union" name:IDENT generics:Generics? where_clause:WhereClause? "{" fields:RecordField* % "," ","? "}" } |
+    Trait:{
+        unsafety:"unsafe"? auto:"auto"? "trait" name:IDENT generics:Generics?
+        { ":" superbounds:TypeBound* % "+" "+"? }?
+        where_clause:WhereClause? "{" trait_items:TraitItem* "}"
+    } |
+    TraitAlias:{
+        "trait" name:IDENT generics:Generics?
+        { ":" superbounds:TypeBound* % "+" "+"? }?
+        where_clause:WhereClause? "=" bounds:TypeBound* % "+" "+"? ";"
+    } |
+    Impl:{
+        defaultness:"default"? unsafety:"unsafe"? "impl" generics:Generics?
+        { negate:"!"? trait_path:Path "for" }? ty:Type
+        where_clause:WhereClause? "{" attrs:InnerAttr* impl_items:ImplItem* "}"
+    } |
+    Macro:{ "macro" name:IDENT { "(" TOKEN_TREE* ")" }? "{" TOKEN_TREE* "}" } |
+    MacroCall:ItemMacroCall;
+
+UseTree =
+    Glob:{ prefix:UseTreePrefix? "*" } |
+    Nested:{ prefix:UseTreePrefix? "{" children:UseTree* % "," ","? "}" } |
+    Simple:{ path:Path { "as" rename:IDENT }? };
+UseTreePrefix =
+    Path:{ path:Path "::" } |
+    Global:"::";
+
+ForeignItem = attrs:OuterAttr* vis:Vis? kind:ForeignItemKind;
+ForeignItemKind =
+    Static:{ "static" mutable:"mut"? name:IDENT ":" ty:Type ";" } |
+    Fn:{ "fn" decl:FnDecl ";" } |
+    Type:{ "type" name:IDENT ";" } |
+    MacroCall:ItemMacroCall;
+
+TraitItem = attrs:OuterAttr* kind:TraitItemKind;
+TraitItemKind =
+    Const:{ "const" name:IDENT ":" ty:Type { "=" default:Expr }? ";" } |
+    Fn:{ header:FnHeader "fn" decl:FnDecl { default_body:Block | ";" } } |
+    Type:{ "type" name:IDENT generics:Generics? { ":" bounds:TypeBound* % "+" "+"? }? where_clause:WhereClause? { "=" default:Type }? ";" } |
+    MacroCall:ItemMacroCall;
+
+ImplItem = attrs:OuterAttr* defaultness:"default"? vis:Vis? kind:ImplItemKind;
+ImplItemKind =
+    Const:{ "const" name:IDENT ":" ty:Type "=" value:Expr ";" } |
+    Fn:{ header:FnHeader "fn" decl:FnDecl body:Block } |
+    Type:{ "type" name:IDENT generics:Generics? where_clause:WhereClause? "=" ty:Type ";" } |
+    ExistentialType:{ "existential" "type" name:IDENT generics:Generics? where_clause:WhereClause? ":" bounds:TypeBound* % "+" "+"? ";" } |
+    MacroCall:ItemMacroCall;
+
+FnHeader = constness:"const"? unsafety:"unsafe"? asyncness:"async"? { "extern" abi:Abi }?;
+FnDecl = name:IDENT generics:Generics? "(" args:FnArgs? ","? ")" { "->" ret_ty:Type }? where_clause:WhereClause?;
+
+// FIXME(eddyb) find a way to express this `A* B?` pattern better
+FnArgs =
+    Regular:FnArg+ % "," |
+    Variadic:"..." |
+    RegularAndVariadic:{ args:FnArg+ % "," "," "..." };
+FnArg =
+    SelfValue:{ mutable:"mut"? "self" } |
```

Comment:
This misses type ascription I think (which is stable in limited form)

```
struct S;
impl S {
    fn f(self: S) {}
}
fn main() {}
```

Reply:
Note that that parses, but `self` there is just a binding pattern.

Reply:
@Centril do we yet have a proposal for `"self" "@" Pat { ":" Type }?`, or were there previous attempts that died out? IMO it would be a natural extension.